mypy/meet.py

```diff
@@ -138,9 +138,11 @@ class C(A, B): ...
 
     # We must check for TupleTypes before Instances, since Tuple[A, ...]
     # is an Instance
-    tup_overlap = is_overlapping_tuples(t, s, use_promotions)
-    if tup_overlap is not None:
-        return tup_overlap
+    if is_overlapping_tuples(t, s, use_promotions):
+        return True
+
+    if is_subtype(t, s) or is_subtype(s, t):
+            return True
```

Comment:
On one hand I would like to see this general case somewhere at the top of this function as an obvious case of overlap. However the fact that such an obvious case is still not here makes me think of "Chesterton's fence". Is there any danger of putting this here (like performance costs)?

An alternative may be to put `return is_subtype(t, s) or is_subtype(s, t)` instead of `return False` near the end of `is_overlapping_tuples`.

If you think the performance is still fine, then I would keep this check here (or maybe even higher above), but add few more tests that cover the other (non-tuple) situations not covered by existing `if` branches.

Reply:
Yeah, I agree performance is something we might need to think about here, since we're basically traversing the types three times instead of once now. I think the subtype cache might help mitigate this, though of course it's hard to say for certain w/o benchmarking.

I suspect the other reason why the function doesn't already do something like this was because it was originally intended to be super-simple/lightweight -- here's a [snapshot from 2015](https://github.com/python/mypy/blob/081532d1e291206cdf6df7db1c68d4a8667f4adc/mypy/meet.py#L50) and [from 2016](https://github.com/python/mypy/blob/e308f2f786d432577286e0d84abaca505d02ec81/mypy/meet.py#L39), for example. I guess it then just steadily accumulated patches over the years?

Reply:
I don't particularly have a great idea of how I could benchmark this, but I am happy to move it to the end if you think that is better.

Reply:
> I don't particularly have a great idea of how I could benchmark this

A good benchmark is self checking `mypy mypy`.